type_decoders/README.md

```diff
@@ -0,0 +1,787 @@
+
+if (result instanceof DecoderSuccess) {
+  const value: Date = result.value;
+}
+```
+
+This decoder will verify that a string is in a `YYYY-MM-DD` format and, if so, convert the string to a date. The return type of the decoder is `Date`.
+
+## Decoder API
+
+### assert()
+
+```ts
+export function assert<R, V>(decoder: Decoder<R, V>): { (value: V): R; (value: Promise<V>): Promise<R> };
+export function assert<R, V>(decoder: PromiseDecoder<R, V>): (value: V | Promise<V>) => Promise<R>;
+```
+
+`assert()` accepts a single decoder as an argument and returns a new function which can be used to decode the same values as provided decoder. On decode success, the validated value is returned directly and on failure the `DecoderError` is thrown (rather than returned).
+
+Example:
+
+```ts
+const validator = assert(isNumber());
+
+const value: number = validator(1);
+
+const value: number = validator('1'); // will throw a `DecoderError`
+```
+
+### isBoolean()
+
+```ts
+interface IBooleanDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isBoolean(options?: IBooleanDecoderOptions): Decoder<boolean, unknown>;
+```
+
+`isBoolean()` can be used to verify that an unknown value is a `boolean`.
+
+### isString()
+
+```ts
+interface IStringDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isString(options?: IStringDecoderOptions): Decoder<string, unknown>;
+```
+
+`isString()` can be used to verify that an unknown value is a `string`.
+
+### isNumber()
+
+```ts
+interface INumberDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isNumber(options?: INumberDecoderOptions): Decoder<number, unknown>;
+```
+
+`isNumber()` can be used to verify that an unknown value is a `number`.
+
+### isInteger()
+
+```ts
+interface IIntegerDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isInteger(options?: IIntegerDecoderOptions): Decoder<number, unknown>;
+```
+
+`isInteger()` can be used to verify that an unknown value is a whole `number`.
+
+### isUndefined()
+
+```ts
+interface IUndefinedDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isUndefined(options?: IUndefinedDecoderOptions): Decoder<undefined, unknown>;
+```
+
+`isUndefined()` can be used to verify that an unknown value is `undefined`. This is a convenience function for `isExactly(undefined)`.
+
+### isNull()
+
+```ts
+interface INullDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isNull(options?: INullDecoderOptions): Decoder<null, unknown>;
+```
+
+`isNull()` can be used to verify that an unknown value is `null`. This is a convenience function for `isExactly(null)`.
+
+### isRegex()
+
+```ts
+interface IRegexDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isRegex(regex: RegExp, options?: IRegexDecoderOptions): Decoder<string, unknown>;
+```
+
+`isRegex()` can be used to verify that an unknown value is `string` which conforms to the given `RegExp`.
```

Comment:
I’m not exactly sure how much you’re allowing matchers to overlap, but would it make sense to re-type this as taking in a string and outputting a string, leaving the actual check to see if the input is a string to `isString`? Or would that be too much DRY?

Reply:
It's a good question. This *might* (not sure) be the only decoder which overlaps in this way and I did think about it a bit. Ultimately, I feel like that would be too dry. For 1) I think most uses of this decoder will be to simply check that something is a string of a certain format. Requiring someone to manually compose isChainOf([isString(), isRegex()]) all the time would be annoying. Possibly more importantly, currently all the exported decoders in the module expect `unknown` inputs. At the moment, I think there is value in maintaining that consistency, even if it results in the occasional, unnecessary, `isString()` check. I.e. predictability as a feature.